File: display/lua/geometry.lua
local geometry = {}

function geometry.close(a, b, tolerance)
  return math.abs(a - b) <= (tolerance or 8)
end

function geometry.sameFrame(a, b, tolerance)
  return geometry.close(a.x, b.x, tolerance)
    and geometry.close(a.y, b.y, tolerance)
    and geometry.close(a.w, b.w, tolerance)
    and geometry.close(a.h, b.h, tolerance)
end

function geometry.center(rect)
  return { x = rect.x + rect.w / 2, y = rect.y + rect.h / 2 }
end

function geometry.dist2(a, b)
  local dx, dy = a.x - b.x, a.y - b.y
  return dx * dx + dy * dy
end

function geometry.clamp(value, minValue, maxValue)
  if value < minValue then return minValue end
  if value > maxValue then return maxValue end
  return value
end

function geometry.halves(screen)
  local sf = screen:frame()
  return {
    left = hs.geometry.rect(sf.x, sf.y, sf.w / 2, sf.h),
    right = hs.geometry.rect(sf.x + sf.w / 2, sf.y, sf.w / 2, sf.h),
  }
end

function geometry.thirds(screen)
  local sf = screen:frame()
  return {
    leftTwoThirds = hs.geometry.rect(sf.x, sf.y, sf.w * 2 / 3, sf.h),
    rightOneThird = hs.geometry.rect(sf.x + sf.w * 2 / 3, sf.y, sf.w / 3, sf.h),
  }
end

function geometry.quads(screen)
  local sf = screen:frame()
  local hw, hh = sf.w / 2, sf.h / 2
  return {
    hs.geometry.rect(sf.x,      sf.y,      hw, hh),
    hs.geometry.rect(sf.x + hw, sf.y,      hw, hh),
    hs.geometry.rect(sf.x,      sf.y + hh, hw, hh),
    hs.geometry.rect(sf.x + hw, sf.y + hh, hw, hh),
  }
end

function geometry.centered(screen, scale)
  local sf = screen:frame()
  return hs.geometry.rect(
    sf.x + sf.w * (1 - scale) / 2,
    sf.y + sf.h * (1 - scale) / 2,
    sf.w * scale,
    sf.h * scale
  )
end

return geometry

File: display/src/displayctl/geometry.py
from __future__ import annotations


TOLERANCE = 8.0


def close(a: float, b: float, tolerance: float = TOLERANCE) -> bool:
    return abs(a - b) <= tolerance


def same_frame(a: dict[str, float], b: dict[str, float], tolerance: float = TOLERANCE) -> bool:
    return (
        close(a["x"], b["x"], tolerance)
        and close(a["y"], b["y"], tolerance)
        and close(a["w"], b["w"], tolerance)
        and close(a["h"], b["h"], tolerance)
    )


def halves(x: float, y: float, w: float, h: float) -> tuple[dict[str, float], dict[str, float]]:
    return (
        {"x": x, "y": y, "w": w / 2, "h": h},
        {"x": x + w / 2, "y": y, "w": w / 2, "h": h},
    )


def weighted_left(x: float, y: float, w: float, h: float) -> tuple[dict[str, float], dict[str, float]]:
    return (
        {"x": x, "y": y, "w": w * 2 / 3, "h": h},
        {"x": x + w * 2 / 3, "y": y, "w": w / 3, "h": h},
    )


def center(frame: dict[str, float]) -> dict[str, float]:
    return {"x": frame["x"] + frame["w"] / 2, "y": frame["y"] + frame["h"] / 2}


def dist2(a: dict[str, float], b: dict[str, float]) -> float:
    dx = a["x"] - b["x"]
    dy = a["y"] - b["y"]
    return dx * dx + dy * dy


def quads(x: float, y: float, w: float, h: float) -> list[dict[str, float]]:
    return [
        {"x": x, "y": y, "w": w / 2, "h": h / 2},
        {"x": x + w / 2, "y": y, "w": w / 2, "h": h / 2},
        {"x": x, "y": y + h / 2, "w": w / 2, "h": h / 2},
        {"x": x + w / 2, "y": y + h / 2, "w": w / 2, "h": h / 2},
    ]
